lib/model/typing_status.dart

```diff
@@ -0,0 +1,82 @@
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import 'narrow.dart';
+
+/// The model for tracking the typing status organized by narrows.
+///
+/// Listeners are notified when a typist is added or removed from any narrow.
+class TypingStatus extends ChangeNotifier {
+  TypingStatus({
+    required this.selfUserId,
+    required this.typingStartedExpiryPeriod,
+  });
+
+  final int selfUserId;
+  final Duration typingStartedExpiryPeriod;
+
+  Iterable<SendableNarrow> get debugActiveNarrows => _timerMapsByNarrow.keys;
+
+  Iterable<int> typistIdsInNarrow(SendableNarrow narrow) =>
+    _timerMapsByNarrow[narrow]?.keys ?? [];
+
+  // Using SendableNarrow as the key covers the narrows
+  // where typing notifications are supported (topics and DMs).
+  final Map<SendableNarrow, Map<int, Timer>> _timerMapsByNarrow = {};
+
+  @override
+  void dispose() {
+    for (final timersByTypistId in _timerMapsByNarrow.values) {
+      for (final timer in timersByTypistId.values) {
+        timer.cancel();
+      }
+    }
+    super.dispose();
```

Comment:
```suggestion
  void dispose() {
    for (final timersByTypistId in _timerMapsByNarrow.values) {
      for (final timer in timersByTypistId.values) {
        timer.cancel();
      }
    }
    _timerMapsByNarrow.clear();
    super.dispose();
```

In general there isn't a need for `dispose` methods to clear maps and such just for the sake of clearing them; it's the garbage collector's job to discard them when the object is truly discarded.

But in this case, since we're disposing the timers, let's also clear the map, because that way this maintains the basic invariant of this data structure: all the timers in the map are live, and all the live timers this class created are in the map.